src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ADRController.kt

```diff
@@ -201,20 +205,48 @@ class ADRController(private val encryption: Encryption,
 
         // 4. Checksum file and upload with metadata to ADR
         val filePart = Pair("upload", file)
+        val fileHash = file.md5sum()
         val commonParameters =
-                listOf("name" to resourceFileName, "description" to artefact.second, "hash" to file.md5sum(),
+                listOf("name" to resourceFileName, "description" to artefact.second, "hash" to fileHash,
                         "resource_type" to resourceType)
         val adr = adrClientBuilder.build()
         return try
         {
             when (resourceId)
             {
                 null -> adr.postFile("resource_create", commonParameters + listOf("package_id" to id), filePart)
-                else -> adr.postFile("resource_patch", commonParameters + listOf("id" to resourceId), filePart)
+                else ->
+                {
+                    if (uploadFileHasChanges(resourceId, fileHash))
+                    {
+                        adr.postFile("resource_patch", commonParameters + listOf("id" to resourceId), filePart)
+                    }
+                    else
+                    {
+                        EmptySuccessResponse.asResponseEntity()
+                    }
+                }
             }
         }
-        finally {
+        catch (e: Exception)
+        {
+            ErrorDetail(HttpStatus.INTERNAL_SERVER_ERROR, e.message!!).toResponseEntity()
+        } finally
+        {
             tmpDir.deleteRecursively()
         }
     }
+
+    fun uploadFileHasChanges(resourceId: String, newDatasetHash: String): Boolean
+    {
+        val adr = adrClientBuilder.build()
+        val response = adr.get("resource_show?id=${resourceId}")
+        if (response.statusCode.isError) {
+            throw IOException("Unable to retrieve hash")
```

Comment:
This is looking good now. If this error message is going to make it back to the user (which I think it will), would suggest possibly being a little more descriptive here, if only to add "from ADR" to give some idea of what's gone wrong. 

And I'd still vote for using objectMapper!